### ISSUE TITLE
Enhancement or bug?: report "nested" validation errors in cases of oneOf/anyOf

### ISSUE THREAD

**Issue by nickivanov** [tdegrunt/jsonschema#189]
Currently validations of `oneOf`/`anyOf`/`allOf` return simply a binary state indicating that the validation failed. However, the root causes of one/some/all of nested schemas failing cannot be determined. In a modestly complex schema that makes use of definitions for common groups of elements it's very hard to troubleshoot validation errors if all you get is "there's an error there somewhere".

It would be very useful if there were a global array or stack of all validation errors encountered by the validator, which `Validator.prototype.validate()` would return. 

Or may be it's a bug that `Validator.prototype.validateSchema()` creates its local instance of `ValidationResult` and it should use the global one instead?

**Comment by awwright** [comment]
Can you provide some sample output?

`allOf` should pass up and show errors from all sub-schemas.

Unfortunately for the other keywords, there's no single way to determine what needs correction in order to produce a valid schema. `allOf` and `anyOf` both produce situations where there's multiple values you could fix, and you need to change just one value without touching the rest of the document, to fix validation.

**Comment by nickivanov** [comment]
@awwright you're right, `allOf` does return nested validation failures; `anyOf` and `oneOf` do not. I think it would be OK if in the latter two cases multiple errors were returned; it would still be better than knowing no details at all.

**Comment by awwright** [comment]
Hmm, lemme examine what other implementations are doing. Right now the errors lists everything that _must_ be fixed.

Changing this for "anyOf" and "oneOf" would mean that it would cause an error to fix all the errors that are returned.

Or maybe I could hide this behind an "options" flag?

**Comment by nickivanov** [comment]
I understand the concern. The validator cannot really tell _which_ of the failed legs must be fixed, because it doesn't know the semantics of the document it validates. The use case I have deals with the schema describing a sort of event log record; there may be different types of events, they all have a common "header", but the rest of the document differs between various event types. Basically it looks like this:
```
$schema: ...
oneOf:
 - ref: #/event_type1
 - ref: #/event_type2
 ...

event_type1:
allOf:
 - ref: #/event_common_fields
 - properties:
  event_type:
   type: string
   pattern: "EVT1"
...
```
etc. So, when validating a certain event record, it should fail against all event types but one. If it fails against all of them, I need to know which specific field is invalid. Looking at the full stack of errors I can see which failures are expected (because of the wrong event_type value) and which one is not.

I think providing an "options" flag to enable this extra reporting is a good idea. How about "options.report_nested_errors"?

**Comment by nickivanov** [comment]
Looks like I'm not alone and this is a long-standing request (see #98). How about we actually fix this?

**Comment by tdegrunt** [comment]
So, I'm not entirely up to speed with this issue. But: Is there a pull request which actually fixes this issue? If yes, then let's accept that pull request, if not: we will happily accept pull requests. 
Please point me to the pull request for the first case.

**Comment by nickivanov** [comment]
There's one from 2015 ( #98 ) and there's the one I just created before I saw the other one ( #190 ) -- they do the same thing with minor differences, so you can pick either. In favour of #190, I also updated the readme :)

**Comment by ashleyryan** [comment]
Is there any plan to merge this pull request in? I'm looking for the same feature.

**Comment by tdegrunt** [comment]
Just merged #190. Closing this issue.
